### ISSUE TITLE
The diagnostic_table method gives inconsistent results in presence of behavior responses

### ISSUE THREAD

**Issue by GoFroggyRun** [open-source-economics/Tax-Calculator#1845]
This issue is related to #1840, #1843 and #1827. 

My question is that how should behavioral assumptions be implemented inTC. I used two different ways to implement behavioral assumptions (with the reform being `2017_law.json`) and got inconsistent results. In particular, the choices of years where calculators choose to advance to, via `advance_to_year` option, seem to affect the output. 

I used the following script to do a simple test:
```
import pandas as pd
import numpy as np
import sys
sys.path.append("../..")
import taxcalc
from taxcalc import Policy, Records, Calculator, Behavior

def diag_table_under_beh(year):
    assump_json = """
        {"behavior": {"_BE_sub": {"2018": [0.25]}},
         "growdiff_baseline": {},
         "growdiff_response": {},
         "consumption": {}
        }
    """
    params = Calculator.read_json_param_objects("../reforms/2017_law.json", assump_json)

    puf_fullsample = pd.read_csv("../../puf.csv")

    policy_base = Policy()
    records_base = Records(data=puf_fullsample)

    policy_reform = Policy()
    records_reform = Records(data=puf_fullsample)
    policy_reform.implement_reform(params['policy'])

    behv0 = Behavior()
    behv0.update_behavior(params['behavior'])

    calcbase = Calculator(policy = policy_base, records = records_base )
    calcreform = Calculator(policy = policy_reform, records = records_reform, behavior=behv0)

    calcbase.advance_to_year(year)
    calcreform.advance_to_year(year)
    calcreform = Behavior.response(calcbase, calcreform)

    return calcreform.diagnostic_table(num_years=(2026 - year))

tab_2018 = diag_table_under_beh(2018)
tab_2020 = diag_table_under_beh(2020)

if not (tab_2018[2020]['Combined Liability ($b)'] == tab_2020[2020]['Combined Liability ($b)']):
    print(tab_2018[2020]['Combined Liability ($b)'])
    print(tab_2020[2020]['Combined Liability ($b)'])
```
and got the following result when executed:
```
(taxcalc-dev) Sean-Wangs-MBP:tests seanwang$ python behav.py 
You loaded data for 2009.
Tax-Calculator startup automatically extrapolated your data to 2013.
You loaded data for 2009.
Tax-Calculator startup automatically extrapolated your data to 2013.
You loaded data for 2009.
Tax-Calculator startup automatically extrapolated your data to 2013.
You loaded data for 2009.
Tax-Calculator startup automatically extrapolated your data to 2013.
3084.49789598
3651.45265953
```

The script intends to investigate the aggregate combined liability in 2020 under reform `2017_law.json` and assumption `"_BE_sub": {"2018": [0.25]}`. The results seem to be very different when calculators were advanced to different years, namely 2018 and 2020 in this example. In other words, the behavioral mechanism in tax-calculator seems to be year-dependent (the year where calculators are at). 

And I am wondering which way is the right one to implement behavioral assumptions? @martinholmer @MattHJensen @Amy-Xu

**Comment by martinholmer** [comment]
Sean, this is much, much better.  Thank you.

I agree with @GoFroggyRun in issue #1845 that his free-standing script has found evidence of a bug in the Tax-Calculator library.  I think any user of the Calculator `diagnostic_table` method would expect the same 2020 results to be generated by the two different calls to `diagnostic_table` that Sean makes in his script.

So, Sean, as I [said a few days ago](https://github.com/open-source-economics/Tax-Calculator/issues/1827#issuecomment-360257239), this is how you should proceed:
> we have a standard procedure to deal with bugs. You should prepare a two-part pull request that does this:
>
> 1. **Identify bug** by creating a new test that fails because of the buggy logic
>
> 2. **Correct bug** by fixing the buggy logic so that the new test now passes and all old tests continue to pass

Let me offer some suggestions about creating the new test in step 1.  Make the new test as simple as possible so that it has no dependencies, runs as fast as possible, and uses CPS data rather PUF data so that it will always run on GitHub.  In particular this means seeing if you get the same test failure with a simpler reform that is specified in the code of the new test, and seeing if you get the same test failure using the cps_subsample fixture.

Simplifying the reform is especially important because you don't want the meaning of the test to change just because somebody changes the `taxcalc/reforms/2017_law.json` file.  And simplifying the reform may help you find the bug.  Remember that, so far, you have evidence of a bug, but presumably don't yet know what code is buggy, and therefore, causing the test failure.

As you go about your work in dealing with this bug, think about what a doctor does when a patient says "I'm sick".  First, the doctor asks the patient what the symptoms are (fever, rash, nausea, etc.?)  Then given the reported symptoms, the doctor makes a **diagnosis**: that is, the doctor identifies the illness causing the symptoms and then decides on a **treatment** to cure the illness without causing any harmful side-effects.

As you simplify the test, keep thinking about the range of possible diagnoses.  Over-simplifying the test in step 1 (so that it passes) produces valuable diagnostic information because whatever simplification caused the test to pass provides information about what diagnosis to mark off you list.  This whole process of diagnosis can be the most difficult part of resolving a bug.  Once you have a correct diagnosis, the treatment (bug fix) is usually relatively easy.  And, unlike doctors, you have an easy way to tell that your treatment cures the illness (the new test passes) and causes no harmful side-effects (all the old test continue to pass).

**Comment by GoFroggyRun** [comment]
Thanks for your vivid, analogous example @martinholmer. Thanks very much for your fix in #1849. 
Closing in favor of #1849.
